core/domain/html_validation_service.py

```diff
@@ -887,23 +887,28 @@ def get_invalid_svg_tags_and_attrs(svg_string):
         svg_string: str. The SVG string.
 
     Returns:
-        tuple. A 2-tuple, the first element of which is a list of invalid tags,
-        and the second element of which is a list of invalid attributes.
+        tuple(list(str), list(str)). A 2-tuple, the first element of which
+        is a list of invalid tags, and the second element of which is a
+        list of invalid tag-specific attributes.
+        eg. (['invalid-tag1', 'invalid-tag2'], ['path:invalid-attr'])
```

Comment:
Explain clearly, in the description, the format of each entry in the second list in the tuple.

Reply:
Thumbnail creators would be using software like Photoshop/Illustrator etc to create these svg files. I am worried that the creators would not easily be able to edit the svg (without affecting the image) in case the generated svg contains invalid tags. Also different svg generation softwares may use different tags or attributes to create the image. So we may end up having to add more and more tags to the whitelist.
Given this, Is it worth creating a blacklist rather than a whitelist?

Reply:
We have such a strict validation to prevent security issues from creeping in. I think the general standard in security is to whitelist user input rather than relying on a blacklist. Most of the popular HTML sanitizers like [html5lib](https://blog.whatwg.org/tag/html5lib) (which python bleach uses) and AngularJS [$sanitize](https://github.com/angular/angular.js/blob/master/src/ngSanitize/sanitize.js#L384) use this technique. We used the same master list that these projects use to create our own whitelist -- so it should be pretty standard. That said, I understand your concern regarding users not being able to upload / use an image they want because of these restrictions. But, as I recently found out, softwares like Adobe Illustrator provide a way to export SVGs such that it is safe to use on the web (see [this](https://helpx.adobe.com/in/illustrator/using/svg.html#enhancements)). I tested some images that were created in this manner and they seemed to work as expected. We would be documenting some guidelines for creating thumbnails and maybe that could include links to software specific config / troubleshooting there (if this proves to be a real issue). What do you think?

Reply:
That sounds great! The Adobe link you shared is very useful. I am guessing you would add the guidelines in our wiki? If so, whenever the svg gets rejected for any invalid tags, it might be worth mentioning that we have some guidelines (and link to it) as part of the error response to the creator.